### PR TITLE
add ssl traffic support

### PR DESCRIPTION
**Pull request by klay-ke**

### CODE REVIEW

File: client/src/main/java/com/vesoft/nebula/client/graph/data/CASignedSSLParam.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
+/* Copyright (c) 2020 vesoft inc. All rights reserved.
```
Comment (Nicole00) [inline client/src/main/java/com/vesoft/nebula/client/graph/data/CASignedSSLParam.java:1]: Copyright (c) 2021

File: client/src/main/java/com/vesoft/nebula/util/SslUtil.java [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+            JcaX509CertificateConverter certificateConverter =
+                    new JcaX509CertificateConverter().setProvider("BC");
+
+            // Load client certificate
+            reader = new PEMParser(new FileReader(crtFile));
+            X509CertificateHolder certHolder = (X509CertificateHolder) reader.readObject();
+            reader.close();
+
+            X509Certificate cert = certificateConverter.getCertificate(certHolder);
+            // CA certificate is used to authenticate server
```
Comment (Aiee) [inline client/src/main/java/com/vesoft/nebula/util/SslUtil.java:145]: client cert

**Review by Aiee — COMMENTED** [review]
Consider using environment variable in docker-compose to prevent using multiple yaml files.

File: client/src/main/java/com/vesoft/nebula/client/graph/data/CASignedSSLParam.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+ */
+
+package com.vesoft.nebula.client.graph.data;
+
+public class CASignedSSLParam extends SSLParam {
+    private String caCrtFilePath;
+    private String crtFilePath;
+    private String keyFilePath;
+
+    public CASignedSSLParam(String caCrtFilePath, String crtFilePath, String keyFilePath) {
```
[the remark this answers is not in the document]
Comment (klay-ke) [reply client/src/main/java/com/vesoft/nebula/client/graph/data/CASignedSSLParam.java:14]: all these files are required to send a SSL CA signed traffic

File: client/src/main/java/com/vesoft/nebula/client/graph/net/SyncConnection.java [hunk 4]
```diff
@@ -143,6 +192,7 @@ public boolean ping() {
             execute(0, "YIELD 1;");
             return true;
         } catch (IOErrorException e) {
+            e.printStackTrace();
```
Comment (Nicole00) [inline client/src/main/java/com/vesoft/nebula/client/graph/net/SyncConnection.java:195]: use log to print the stac.

File: client/src/main/java/com/vesoft/nebula/util/SslUtil.java [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+            // Create SSL socket factory
+            SSLContext context = SSLContext.getInstance("TLSv1.3");
+            context.init(keyManagerFactory.getKeyManagers(),
+                    trustManagerFactory.getTrustManagers(), null);
+
+            // Return the newly created socket factory object
+            return context.getSocketFactory();
+
+        } catch (Exception e) {
+            e.printStackTrace();
```
Comment (Nicole00) [inline client/src/main/java/com/vesoft/nebula/util/SslUtil.java:103]: ditto

File: client/src/test/resources/ssl/selfsigned.key [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+l+tGi9MeSBQeWKV6D3AVnO05AZjfTUzSK0vw4DgNh5YPNJvLy31B7kDAS88vyGI1
+t6MBwjW4/tz/nS/p1Go3mSzBhPkIsCrZE+ar7lH8p8JqkLl4fXIMaVKIfyfJdzyS
+lkh3K7bOGDPegxxxaWdb+EnC7k+1R3EOU7uJFW61HyrGI3q6Y7kOl5aYSJ5Ge1Uv
+PycFWHWVTHq/R7HRE6HIJzGe/PnLIbStXLDFeivjfcYq1YaSaF8Vl+xg+0u3ULOl
+P6IuPTph6dlcgttRZVl3ETcF0T+2wfbUwgjf0ZiguCJfR2jLGhPl1KBg0Kd9cTSY
+zI3YMMd2G8hApt/QFlm4Ry8CqaJUmDcjDNIJT3M+RldUgfz37NsX05cA5e9+I1AL
+2406F/v5U9gWsYx7HuwJtQrDzYYDbl1GD4H+qHFJE5JYhPP4AyWYxJ1NR5dqyvrt
++3r5+xlwZrS76c10RsBWL7th8ZEzRxOZxbtLwbf4bG/tIGfQP2sTnWwA+qym6b2S
+sRduqOTP+xwnhOq/ZKn8lfsDfhT8CPnKHBsd09kM9y/UWuxFe0upLydRLE/Wsb9s
+-----END RSA PRIVATE KEY-----
```
Comment (Nicole00) [inline client/src/test/resources/ssl/selfsigned.key:30]: add extra empty line at the end.

File: client/src/test/resources/ssl/selfsigned.password [hunk 7]
```diff
@@ -0,0 +1 @@
+vesoft
```
Comment (Nicole00) [inline client/src/test/resources/ssl/selfsigned.password:1]: ditto

File: examples/src/main/java/com/vesoft/nebula/examples/GraphClientExample.java [hunk 8]
```diff
@@ -146,22 +148,58 @@ public static void main(String[] args) {
             {
                 String queryForJson = "YIELD 1";
                 String resp = session.executeJson(queryForJson);
-                JSONObject errors = JSON.parseObject(resp).getJSONArray("result").getJSONObject(0)
-                        .getJSONObject("errors");
-                if (!errors.getString("errorCode").equals("0")) {
+                JSONObject errors = JSON.parseObject(resp).getJSONArray("errors").getJSONObject(0);
+                if (errors.getInteger("code") != 0) {
```
Comment (Nicole00) [inline examples/src/main/java/com/vesoft/nebula/examples/GraphClientExample.java:152]: use ErrorCode.SUCCESS.getErrorCode?

File: client/src/main/java/com/vesoft/nebula/client/graph/net/SyncConnection.java [hunk 9]
```diff
@@ -15,23 +15,68 @@
...
+    private SSLParam sslParam = null;
+    private boolean enabledSsl = false;
+
+    @Override
+    public void open(HostAddress address, int timeout, SSLParam sslParam) throws IOErrorException {
+        try {
+            SSLSocketFactory sslSocketFactory;
+
+            this.sslParam = sslParam;
+            if (sslParam == null) {
```
Comment (HarrisChu) [inline client/src/main/java/com/vesoft/nebula/client/graph/net/SyncConnection.java:45]: why need this? 
if sslParm is null, ConnObjectPool.java would raise IllegalArgumentException

**Pull request closed without merging**